### DISCUSSION TITLE
Control button - can I get it's state?

### DISCUSSION THREAD

**Discussion by YouCanNotBeSerious** [HomeSpan/HomeSpan#732] [Q&A]
Due to hardware constraints, I want to use the controlPin button in my part of the code. 

Reading it with digitalRead() seems wrong when clearly HomeSpan is already monitoring it - so is there a way for my code to know the button status by checking some existing flag?  I know it is a SpanButton type so has SINGLE press, LONG press and so on. RealPushButtons.ino seems to be in the right direction so I am looking  at that right now - like `if(pressType==SpanButton::SINGLE){` 

Also semi - related, a simple option when setting up the status LED to invert it would be wonderful - again, hardware constraints, I have replaced the module in an existing wifi smart switch with an ESP module and the built in LED is active low.

**Comment by YouCanNotBeSerious** [comment]
I'm trying to help myself as well as asking here :-) This could be related to https://github.com/HomeSpan/HomeSpan/discussions/722 so maybe @homonto got an answer that could help me, or the answer here might help them.

**Comment by HomeSpan** [comment]
Can you explain what you are trying to do with the control pin - if you enable it in HomeSpan, you can't use it for anything else.  If you don't enable it, it's a free pin to do anything you'd like.

**Reply by YouCanNotBeSerious** [reply]
Sure -

the hardware I am repurposing is a controllable outlet - i.e. the usual relay on/off control type.   It was a Tuya unit with their custom wifi chip, but like a lot of this class of hardware you can replace the control module with an ESP based one, which I have done.

So now I have ESP32-C3 control of the nice finished hardware package which is a socket, housing, relay, LEDs and button.....so here comes HomeSpan!   BUT the hardware has only one button which is a standard press=low GPIO, which right now works fine when used as the HomeSpan control button.

I have test code working fine where I can toggle the relay via Home app (basically used the workingLED example, changed service to Outlet etc).  Now I'd like to also toggle the relay when the button is pressed (local control of the outlet) and of course update HomeKit with the new status.

My first thought is just to check the button for a press in loop() which from a quick test seems to at least work, but feels inelegant at best and potentially conflicting with HomeSpan's use of it at worst.  Definitely would appreciate suggestions ;-)

**Reply by YouCanNotBeSerious** [reply]
Actually I have more issues than just getting the button state, I need to also update HomeKit of the relay change.  Hopefully the explanation above shows what I am trying to achieve.

This code below WORKS completely for app control, it's the local button control I need help with. Local button right now toggles the relay, but the app never gets a state update when local control is used.

```

#include "HomeSpan.h"         
#define RELAY 3
#define BUTTON 4 // NOTE this is somewhat overloading the use of HomeSpan control button...bad idea?

struct myOutlet : Service::Outlet 
{         
  int relayPin;                              
  SpanCharacteristic *powerOn;                
  
  myOutlet(int relayPin) : Service::Outlet(){     
    powerOn=new Characteristic::On();           // The outlet's On Characteristic 
    this->relayPin=relayPin;                    
    pinMode(relayPin,OUTPUT);                 
  } 
  
  boolean update(){                         
    digitalWrite(relayPin,powerOn->getNewVal());  
    WEBLOG("Set GPIO %d to %s",relayPin,powerOn->getNewVal()?"ON":"OFF");   
    return(true);                           
  } 
};

void setup() {
  Serial.begin(115200);

  homeSpan.setLogLevel(1);
  homeSpan.enableOTA();
  homeSpan.setControlPin(4);    // GPIO 4 = button
  homeSpan.setStatusPin(5);     // GPIO 5 - LED - note this is assumed high=ON so is inverted on the real hardware
  homeSpan.enableWebLog(10,"pool.ntp.org","UTC","log");           // creates a web log on the URL /HomeSpan-[DEVICE-ID].local:[TCP-PORT]/myLog

  homeSpan.begin(Category::Outlets,"HomeSpan Outlet", "HomeSpan-Outlet1");   // ID for easier OTA updates 
  
  new SpanAccessory();           
    new Service::AccessoryInformation();             
      new Characteristic::Identify();    
    new myOutlet(RELAY);                        
      new Characteristic::OutletInUse(1);   // Usually means something plugged in - noted that the Eve socket has a microswitch to detect this. I am just permanently setting it to true as it is a required characteristic. 
}                  

void loop() {
  homeSpan.poll();
// for local control, toggle relay on button press and update homekit states
if (digitalRead(BUTTON) == LOW)
  {
    digitalWrite(RELAY,!digitalRead(RELAY));
    // update the new relay status to homekit here somehow ???  this part works and toggles the relay but of course Home doesn't know about the state change.
    delay(500);     //debounce for now
  }

}
```

**Reply by YouCanNotBeSerious** [reply]
I'm really a hacker, not a pro at this and I've got this far with a lot of help but I really don't understand how to access the state variables like powerOn that are inside the struct, from my loop().  

It feels like I just need to set powerOn with something like powerOn->setVal(1) where my comment says "    // update the new relay status to homekit here somehow  " because everything else works!

**Reply by YouCanNotBeSerious** [reply]
Replying to myself for other readers to keep the full story flow - loop() shouldn't have any code in it apart from homespan.poll() unless maybe you are a real expert, which I'm not. See below, a button should be added to the myOutlet definition, it happens that it is the same pin used for the HomeSpan control button but that's OK if we only use the single press aspect of the button since HomeSpan looks for a long press to start config mode, and in config mode ignores my use of the button.

**Reply by homonto** [reply]
lets agree to ... disagree - my loop is quite bigger:

```
void loop()
{
  homeSpan.poll();

  #if (USE_MQTT == 1)
    if ((first_heartbeat) and (wifi_connected()) and (mqtt_client.connected()))
    {
      first_heartbeat = false;
      heartbeat();
    }
  #endif
  

  #if (USE_INFLUXDB == 1)
    // check influxdb before first routine - only once
    if (!first_influxdb_check_done)
    {
      // don't print anything here
      // LOG0("[%s]: Checking first time connection with InfluxDB...\n",__func__);
      if (wifi_connected())
      {
        LOG0("[%s]: WiFi OK, Checking first time connection with InfluxDB...\n",__func__);
        if (influxdb_connected())
        {
          first_influxdb_check_done = true;
          WEBLOG("[%s]: Connected to InfluxDB\n",__func__);
          // initial heartbeat - only 1 time
          heartbeat();
        } else 
        {
          LOG0("[%s]: No connection with InfluxDB...\n",__func__);
        }
      } else
      {
        // don't print anything here
        // LOG0("[%s]: WiFi NOT OK\n",__func__);
      }
    }
  #endif

  #if (USE_MQTT == 1)
    if (wifi_connected())
    {
      if (!mqtt_client.connected()) 
      {
        if (   millis() >= (mqtt_reconnect_last_check + (MQTT_CHECK_INTERVAL_S * 1000))   )
        {
           mqtt_reconnect();
           mqtt_reconnect_last_check = millis();
        }
      }
      mqtt_client.loop();
    }
  #endif

  // heartbeat from bridge
  if (millis() >= bridge_heartbeat_interval + (HEARTBEAT_INTERVAL_S * 1000) )
  {
    heartbeat();
    bridge_heartbeat_interval = millis();
  }

  // periodic FW check on server 
  if (millis() >= fw_check_interval + (FW_CHECK_INTERVAL_M * 1000 * 60) )
  {
    periodic_firmware_check();
    fw_check_interval = millis();
  }

}

```

**Reply by YouCanNotBeSerious** [reply]
I did say 

>unless maybe you are a real expert, which I'm not.

I found some of your posts while searching for similar to mine, I think you are quite a bit more expert than me my friend :-)

I am frankly amazed that I sit here with multiple little project boards that talk to HomeKit and are in various stages of working - I am NOT a programmer, I hack my way through, have no formal training or experience, but projects like HomeSpan give such a head start (when I remember to carefully check the examples!). Standing on the shoulders of giants :-)

**Reply by homonto** [reply]
not really but I like to ... try ;-)
I am still waiting for @HomeSpan  to let me (now: us) know how to get the values of some INTERNAL methods/variables inside MY part of the code but it seems I will wait for a bit longer ;-)
still: his job here is AMAZING

**Reply by homonto** [reply]
@YouCanNotBeSerious you can see my full code here:
hub: https://gitlab.papio.uk/zyghom/001-homekit_hub
sensors: https://gitlab.papio.uk/zyghom/001-homekit_sensors

it is a hub that provides data to HomeKit while sensors are sleeping devices around the house, that measure temperature, humidity and light on periodic basis and then sleep again - with 5min sleep time the devices can work for around 1 year on 1 charge (500mAh small LiPo).
Since HomeKit is very limited, I also added sending data to MQTT server so that one can use this data on i.e. Grafana or so to see the long term trends:
<img width="1650" alt="Screenshot 2024-01-07 at 07 25 11" src="https://github.com/HomeSpan/HomeSpan/assets/46562447/83aee979-35be-41b0-8d64-fded16121e76">

**Reply by HomeSpan** [reply]
> I am still waiting for @HomeSpan to let me (now: us) know how to get the values of some INTERNAL methods/variables inside MY part of the code

Sorry if I missed this question before.  There are numerous ways to "extract" Characteristics, Services, etc. from within a class to use elsewhere.  One way is to simply pass the name of a pointer to Characteristics as an argument to whatever other function you are calling.  However, that doesn't work if you are trying to get the value of a Characteristic from inside the main Arduino `loop()`, since there is no way of passing arguments.

In this case you need to do the same things you would if there are variables you create in the Arduino `setup()` function that you need to use in the `loop()` function --- you need to make them global.

For example, below is the DEV_LED structure from Tutorial Example 5:

```
struct DEV_LED : Service::LightBulb {

  int ledPin;
  SpanCharacteristic *power;
  
  DEV_LED(int ledPin) : Service::LightBulb(){

    power=new Characteristic::On();
    this->ledPin=ledPin;
    pinMode(ledPin,OUTPUT);
    
  }

  boolean update(){            

    digitalWrite(ledPin,power->getNewVal());
   
    return(true);  
  } 
};
```

Note how this structure defines `power` and also creates it (using the `new` command).  Instead, if you move the definition `SpanCharacteristic *power;` outside the class as a global variable, you will be able to access it directly from anywhere in your sketch.

Unfortunately, this only works if you have one LED, since `power` is no longer a variable tied to each instance of the class.  So as an alternative, you can *name* the LED structure itself when you create it, so that you can use it later.  Something like this:

```

#include "HomeSpan.h" 
#include "DEV_LED.h"

DEV_LED *led16;     // create global pointers to the DEV_LED Service
DEV_LED *led17;

void setup() {

  Serial.begin(115200);

  homeSpan.begin(Category::Lighting,"HomeSpan LED");
  
  new SpanAccessory();  
    new Service::AccessoryInformation(); 
      new Characteristic::Identify();                
    led16=new DEV_LED(16);                  // save the name using a global pointer


  new SpanAccessory();   
    new Service::AccessoryInformation();    
      new Characteristic::Identify();                       
    led17=new DEV_LED(17);                  // save the name using a global pointer

}

//////////////////////////////////////

void loop(){
  
  homeSpan.poll();

  if(led16->power->getVal() && led17->power->getVal()){
    Serial.printf("Power overload!  Can't have both LED's on at the same time.  Turning off both LEDs...\n");
    led16->power->setVal(false);
    led17->power->setVal(false);
  }
  
}
```

**Reply by homonto** [reply]
thank you @HomeSpan  but my question in here: https://github.com/HomeSpan/HomeSpan/discussions/722 was not about characteristics of the device but general question - in that example it was about wifi status: `homeSpan.connected` - how can I get such value inside my part of the code

**Reply by HomeSpan** [reply]
okay - I'll address at #722

**Reply by YouCanNotBeSerious** [reply]
@HomeSpan your reply does help me with more learning though, so thank you for taking the time to compose it. It was my flawed understanding above that led to me thinking i needed to access the variables from loop(), as you can see i now have achieved what was needed in the "proper" way with your pushing me to the example :-).

**Comment by HomeSpan** [comment]
Hmmm.  You want to create a pushbutton to control the device.  If only there were a series of tutorials that might cover this exact example... 😄 

And yes, you can use the same pin for Control if you do not try to enable a long-press for anything else.

**Reply by YouCanNotBeSerious** [reply]
Heheh yeah... I must admit any time I try to RTFM I seem to miss the nose in front of my own face.... I think it's due to failing to realise that what I am trying to do is usually not anything special or unique no matter how special my mom told me I was ;-)

Thanks for confirming about the control pin, that matches my testing so far, in fact it works quite well in limited bench testing - short press toggles the relay, and long press still works fine to get into config mode, whereupon homespan takes over the button.

I am looking at 15-RealPushButtons now

**Reply by YouCanNotBeSerious** [reply]
OMFG I am so happy to see this line in the debug terminal:  "button pressed" :-)

Added button pin to myOutlet
```
myOutlet(int relayPin, int buttonPin) : Service::Outlet(){     
```
When setting up the accessory send the button pin as well as the relay pin:
```
   new myOutlet(RELAY,BUTTON);                       
```
Now use the button inside myOutlet:

```
  void button(int pin, int pressType) override { 

    if(pressType==SpanButton::SINGLE)  {                      // do the following on a SINGLE press
      LOG1("button pressed");
    }   
  } 
};

```

**Reply by YouCanNotBeSerious** [reply]
@homespan Now it all works! PERFECT.  Thanks for the help and pushing me to the examples.

Not much code, but a lot of new understanding.
```
 void button(int pin, int pressType) override { // this is a built in function called once a spanbutton is set up
    if(pressType==SpanButton::SINGLE)  {                      // do the following on a SINGLE press
      WEBLOG("Local button single-pressed");
      powerOn->setVal(1-powerOn->getVal());                   // toggle the value of the powerOn Characteristic
      digitalWrite(relayPin,powerOn->getNewVal());            // update the actual pin with new value
    }   
  } 
};
```

Now I am going to apply a hardware fix to the status LED because it is inverted on my hardware. Although I do see your other answers about an inverted status LED using the GenericLED it's too much for my mind right now, but I can chop and solder the LED :-P

**Reply by homonto** [reply]
@YouCanNotBeSerious I have been always told: "button: active LOW, LED/relay: active HIGH" that is kind of intuitive:
- in "normal" state, voltage on button can be ANY, unless you press it and ground it
- in "normal" state, voltage on LED/relay should be NONE, unless you press it and pull HIGH
but I have seen in many projects people invert this logics for some reasons.

**Reply by YouCanNotBeSerious** [reply]
I agree with that voltage level logic, but yes in the wild it does vary.  I will say @homespan if you did add an "invert" option to setStatusPin, it can be more than just an option for hardware that is already the "wrong" way around.....  you see, with my LED inverted as it currently stands, the flashes for states like no wifi, not paired, and config mode all still work, you can easily see the flashes.  But when normal running and paired, the LED is off.....in my mind some people might like to use the "wrong" inverted status because the LED is then quiet unless there is something wrong :-)

**Reply by homonto** [reply]
if you inverted the status LED, yet you want to keep it always ON, I think you can now use "inverted" logic of this code:
`homeSpan.setStatusAutoOff(1); // turn OFF LED - not good as no info about the life of bridge then`

it will (if I am not mistaken) try to turn OFF the LED after 1s but if you have the LED inverted, it will actually turn it ON after 1s.
I did not try ;-)

**Reply by HomeSpan** [reply]
You can use an inverted LED (one that turns on when a pin is set to ground, and off when set high) by using: `homeSpan.setStatusDevice(Blinkable *sDev);`.  If you look this up in the API, you'll see is points to a page called Blinker.md which provides you with the exact code to copy to do what you want to do.

**Reply by YouCanNotBeSerious** [reply]
>You can use an inverted LED ....

Copied, pasted, tested, working :-).

This controllable outlet is completely done, now I move on to a second type I have where the module is also pin compatible replaceable for an ESP32 and this one has a power measurement chip on board which has Ardunio libraries for it, so I am going to try and convert this one to HomeSpan.  In case anyone is interested, I realise the power value will only show in third party apps, but once I can read power inside my code, I am thinking of having a "switch" accessory as part of this outlet, and I can decide in my code to set the switch to on/off based on power being above or below a threshold for a time - in other words the power plug will show if the attached washing machine is running or not :-)
